src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1382 @@
+    }, [&](Value exception) {
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return false;
+    });
+  };
+}
+
+Module::EvaluateCallback Module::newWasmModuleHandler(kj::Array<kj::byte> data) {
+  struct Cache {
+    kj::MutexGuarded<kj::Maybe<v8::CompiledWasmModule>> mutex {};
+  };
+  return [data = kj::mv(data), cache = kj::heap<Cache>()](
+      Lock& js,
+      const Url& specifier,
+      const ModuleNamespace& ns,
+      const CompilationObserver& observer) mutable -> bool {
+    return js.tryCatch([&]() -> bool {
+      js.setAllowEval(true);
+      KJ_DEFER(js.setAllowEval(false));
+
+      // Allow Wasm compilation to spawn a background thread for tier-up, i.e. recompiling
+      // Wasm with optimizations in the background. Otherwise Wasm startup is way too slow.
+      // Until tier-up finishes, requests will be handled using Liftoff-generated code, which
+      // compiles fast but runs slower.
+      AllowV8BackgroundThreadsScope scope;
+
+      {
+        // See if we can use a cached compiled module to speed things up.
+        auto lock = cache->mutex.lockShared();
+        KJ_IF_SOME(compiled, *lock) {
+          auto metrics = observer.onWasmCompilationStart(js.v8Isolate);
+          auto result = JsValue(check(
+            v8::WasmModuleObject::FromCompiledModule(js.v8Isolate, compiled)));
+          return ns.setDefault(js, result);
+        }
+      }
+
+      auto module = jsg::compileWasmModule(js, data, observer);
+      auto lock = cache->mutex.lockExclusive();
+      *lock = module->GetCompiledModule();
+      auto result = JsValue(module);
+      return ns.setDefault(js, result);
+    }, [&](Value exception) {
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return false;
+    });
+  };
+}
+
+Function<void()> Module::compileEvalFunction(
+    Lock& js,
+    kj::StringPtr code,
+    kj::StringPtr name,
+    kj::Maybe<JsObject> compileExtensions,
+    const CompilationObserver& observer) {
+  auto metrics = observer.onCjsStyleCompilationStart(js.v8Isolate, name);
+  v8::ScriptOrigin origin(js.str(name));
+  v8::ScriptCompiler::Source source(js.str(code), origin);
+  auto fn = ([&] {
+    KJ_IF_SOME(ext, compileExtensions) {
+      v8::Local<v8::Object> obj = ext;
+      return check(v8::ScriptCompiler::CompileFunction(
+          js.v8Context(), &source, 0, nullptr,
+          1, &obj));
+    } else {
+      return check(v8::ScriptCompiler::CompileFunction(
+          js.v8Context(), &source, 0, nullptr, 0, nullptr));
+    }
+  })();
+
+  return [ref=js.v8Ref(fn)](Lock& js) mutable {
+    js.withinHandleScope([&] {
+      auto fn = ref.getHandle(js);
+      // Any return value is explicitly ignored.
+      auto context = js.v8Context();
+      auto val = JsValue(check(fn->Call(context, context->Global(), 0, nullptr)));
+      JSG_REQUIRE(!val.isPromise(), TypeError,
+          kj::str("Module evaluation function must not return a promise"));
```

Comment:
Largely just because of compatible implementation with other runtimes. Technically it could but there's nothing that can reasonably be expected to be done with it.